src/client-side-metrics/operation-metrics-collector.ts

```diff
@@ -0,0 +1,337 @@
+        this.firstResponseLatency =
+          endTime.getTime() - this.operationStartTime.getTime();
+      }
+    }
+  }
+
+  /**
+   * Called when an operation completes (successfully or unsuccessfully).
+   * Records operation latencies, retry counts, and connectivity error counts.
+   * @param {string} projectId The id of the project.
+   * @param {OperationOnlyAttributes} info Information about the completed operation.
+   */
+  onOperationComplete(projectId: string, info: OperationOnlyAttributes) {
+    if (
+      this.state ===
+      MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS
+    ) {
+      this.state = MetricsCollectorState.OPERATION_COMPLETE;
+      const endTime = new Date();
+      if (projectId && this.operationStartTime) {
+        const totalTime = endTime.getTime() - this.operationStartTime.getTime();
+        {
+          const operationAttributes = {
+            finalOperationStatus: info.finalOperationStatus,
+            streamingOperation: info.streamingOperation,
+            metricsCollectorData: this.getMetricsCollectorData(),
+            clientName: `nodejs-bigtable/${version}`,
+            projectId,
+          };
+          const metrics = {
+            operationLatency: totalTime,
+            retryCount: this.attemptCount - 1,
+            firstResponseLatency: this.firstResponseLatency ?? undefined,
+          };
+          this.metricsHandlers.forEach(metricsHandler => {
+            if (metricsHandler.onOperationComplete) {
+              metricsHandler.onOperationComplete(metrics, operationAttributes);
+            }
+          });
+        }
+      }
+    } else {
+      console.warn('Invalid state transition attempted');
+    }
+  }
+
+  /**
+   * Called when metadata is received. Extracts server timing information if available.
+   * @param {string} projectId The id of the project.
+   * @param {object} metadata The received metadata.
+   */
+  onMetadataReceived(metadata: {
+    internalRepr: Map<string, Buffer>;
+    options: {};
+  }) {
+    const mappedEntries = new Map(
+      Array.from(metadata.internalRepr.entries(), ([key, value]) => [
+        key,
+        value.toString(),
+      ])
+    );
+    const SERVER_TIMING_REGEX = /.*gfet4t7;\s*dur=(\d+\.?\d*).*/;
+    const SERVER_TIMING_KEY = 'server-timing';
+    const durationValues = mappedEntries.get(SERVER_TIMING_KEY);
+    const matchedDuration = durationValues?.match(SERVER_TIMING_REGEX);
+    if (matchedDuration && matchedDuration[1]) {
+      if (!this.serverTimeRead) {
+        this.serverTimeRead = true;
+        this.serverTime = isNaN(parseInt(matchedDuration[1]))
+          ? null
+          : parseInt(matchedDuration[1]);
+      }
+    } else {
+      this.connectivityErrorCount++;
+    }
+  }
+
+  /**
+   * Called when status information is received. Extracts zone and cluster information.
+   * @param {object} status The received status information.
+   */
+  onStatusReceived(status: {
```

Comment:
I'm a little confused why onStatusReceived is different from onMetadataReceived. Aren't they both parsing metadata fields? (I noticed the argument is called metadata)

Reply:
When you create a stream in Node and make a grpc call, the stream receives `status` and `metadata` events. The shape of the status event received matches the first argument of onStatusReceived method and the shape of the metadata event received matches the first argument of onMetadataReceived.